### DISCUSSION TITLE
How to delete virtual env?

### DISCUSSION THREAD

**Discussion by AgentCosmic** [python-poetry/poetry#3690] [Q&A] [answered]
How do I delete a virtual env? I tried just deleting the file but poetry stopped working after that.

**Comment by andyg0808** [comment]
The latest preview version (which can be installed with `get-poetry.py --preview`; see the readme) supports `poetry env remove`, which will handle removing the virtualenv for you. `remove` expects one argument: the Python version to remove the virtualenv for. In my case, I'm using Python 3.7, so I usually type something like `poetry env remove 3.7`.

If you want to see where the virtualenv is stored, there are various ways to print it out. `poetry env info`, which is also new in the preview version, will print a nice summary of information about the current environment Poetry is using, including the full path to the virtualenv.

**Comment by Djailla** [comment]
On macOS, envs are located here `~/Library/Caches/pypoetry/virtualenvs`

**Comment by danielquinn** [comment]
On Linux, the easiest way to find the virtualenv is to pop into a subshell and run `which python`:

```
$ poetry shell
$ which python
```

**Comment by arnegroskurth** [comment]
I feel like this should be easier. Maybe something like `$ poetry destroy [-f]` for the environment associated with the current working directory?

**Reply by alexfazio** [reply]
> @arnegroskurth As pointed out above you can also do `` rm -rf `poetry env info -p` `` to remoe the virtual env directory depending on what you need to do.

@arnegroskurth solution ↑ works well, but I thought it would be beneficial to suggest another method to delete the current Poetry environment. **If you don't know the env name**, you can use the following commands:

```bash
$ exit
$ poetry env remove $(basename $(poetry env info --path))
```

Here's how it works:

1. `exit`: This command ensures that you are not inside an active Poetry shell, which would prevent the deletion of the environment.

2. `poetry env remove $(basename $(poetry env info --path))`:
   - `poetry env info --path`: This command outputs the path to the current virtual environment associated with the Poetry project.
   - `$(...)`: This is command substitution, which executes the command inside the parentheses and replaces it with the output of that command.
   - `basename`: This command retrieves the name of the virtual environment from the path.
   - `poetry env remove`: This command, followed by the environment name, deletes the specified environment.

By using this command sequence, you can effectively find and remove the current Poetry environment without needing to know its name in advance.

**This is a quick yet imperfect method** that can save time in certain situations. I hope this is helpful! Please feel free to ask if you have any additional questions.

**Comment by abn** [comment, marked as answer]
@arnegroskurth the `poetry env remove <python>` command does this.

As pointed out above you can also do ```rm -rf `poetry env info -p` ``` to remoe the virtual env directory depending on what you need to do.

**Reply by malmiteria** [reply]
I'm not sure why this is marked as the answer, as ```poetry env info -p``` doesn't always give the answer.
You need to be in the poetry shell for it to work, otherwise you might just get an empty string if the virtualenv folder is not in the project folder.

Having to open a shell is the kind of things that makes scripting harder, in Makefiles for example, as it is not possible to write a make command that would open the poetry shell, and keep running inside of it (as far as i know)

**Comment by arnegroskurth** [comment]
@abn Sure but that is more complicated and requires the user to know about the `poetry env *` commands. At least me (and I suspect quite a lot of other users) are only using `poetry {install,update,add,remove}` the vast majority of the time and are happy that everything regarding the venv is abstracted away from me.

So I feel like there should be something to just *level the playing-field* without me having to think about venvs again.

**Comment by abn** [comment]
Sounds to me more like a deficiency in the documentation and/or onboarding guide for a new user to Poetry. That is definitely an area that we are interested in improving on.

At this time, I do not think we want to encourage the addition of multiple commands to do the same task. And, personally, adding something like `poetry destroy` does not really solve the issue, it is yet another command anyway that the user needs to know about. Additionally, it can also be ambiguous as it could also mean "destroy everything poetry related".

**Comment by arnegroskurth** [comment]
@abn I just feel like that the benefit of using poetry over venv+pip is for the user to not have to care about two separate tools for dependency-management and project-isolation anymore. And because I can create and update the environment with top-level poetry commands, I feel like that I should also be able to remove the environment with a top-level command that is much more discoverable (`$ poetry help`) then anything under `$ poetry env`.

**Comment by abn** [comment]
That definitely is **one of** the advantages of using poetry. Although, it is important to note that explicit creation of an environment relies on `poetry env use <python>`. What you are referring to is the implicit creation of the environment by commands that require an existing virtual environment. All commands relating to environment managent are already grouped under `env`.

We appreciate your feedback on this. However, for this matter I do not see a reason to change the current behaviour at this time.

**Comment by davnat** [comment]
> As pointed out above you can also do `` rm -rf `poetry env info -p` `` to remoe the virtual env directory depending on what you need to do.

Please note that the suggested command can be quite dangerous if accidentally run in a directory not handled by poetry.

**Comment by timkofu** [comment]
This worked for me:
- cd into the folder where `pyproject.toml` is
- Run `poetry env list` (this will show you the venv for that project)
- Then run `poetry env remove whatever-WhATeVs-py3.9` to delete it

Running `poetry shell` after will create a new venv, after which running `poetry install` will install all the deps listed in `pyproject.toml`.

**Comment by pavanmr94** [comment]
you can actually just run `poetry env remove python` from the project root folder and it will delete its virtualenv

**Comment by mattsta** [comment]
addendum: just note the `python` parameter is your python executable name (it isn't just an argument saying "delete python in this venv"—it's saying "use this executable name to lookup the venv name").

so, if your default path `python` isn't your venv python (e.g. `/usr/bin/python` is still 2.7, but you develop using `/usr/local/bin/python3`), you need to run `poetry env remove python3` or else you'll get yelled at for not having a `py2.7` venv

the lookup is here: [execute code using the `python` argument to find python version](https://github.com/python-poetry/poetry/blob/bf30ca696aa47bcc42937910d620c95c95c0741e/poetry/utils/env.py#L331-L343) then python version is appended to the venv directory. looks like the code is designed to support multi-python-version venvs for the same project, which makes sense, but caught me off guard with needing to fight system python vs. usable python.

**Comment by cgte** [comment]
If you want to remove all environments for a given project: 
`poetry env list --full-path | cut -d ' ' -f 1 | xargs rm -fr `

**Comment by SH2282000** [comment]


**Reply by mattsta** [reply]
The best way continues to be using the built-in way: `poetry env remove python3` and it will _do the right thing_:

```haskell
$ poetry init
$ # ...init things...
$ poetry add pandas
Creating virtualenv test-8DRKBHXr-py3.10 in /Users/matt/Library/Caches/pypoetry/virtualenvs
Using version ^1.5.3 for pandas

Updating dependencies
Resolving dependencies... (0.4s)

Writing lock file

Package operations: 5 installs, 0 updates, 0 removals

  • Installing six (1.16.0)
  • Installing numpy (1.24.2)
  • Installing python-dateutil (2.8.2)
  • Installing pytz (2022.7.1)
  • Installing pandas (1.5.3)
$ poetry env list
test-8DRKBHXr-py3.10 (Activated)
$ poetry env remove python3
Deleted virtualenv: /Users/matt/Library/Caches/pypoetry/virtualenvs/test-8DRKBHXr-py3.10
$ poetry env list
$ # ...nothing there anymore!
$ ls /Users/matt/Library/Caches/pypoetry/virtualenvs/test-8DRKBHXr-py3.10
gls: cannot access '/Users/matt/Library/Caches/pypoetry/virtualenvs/test-8DRKBHXr-py3.10': No such file or directory
```

Entering a venv shell is anti-pattern and should only be used for expert troubheshooting operations.

**Comment by JoaquimEsteves** [comment]
One thing I'm still unclear on is how to remove old environments.

Imagine that in some python project the python-version has been upgraded to from python 3.X to 3.Y

I now want to remove the old one(s) safely and have no earthly idea which ones I should remove.
I'll probably have a few 3.X versions lying around from other projects.

My go to solution has been to periodically remove everything in the `.cache/pypoetry` dir. Which isn't the cleanest

**Comment by maple3142** [comment]
It would be good to have a way to remove env for projects that you have already deleted from the file system. Maybe just a `poetry env remove --all` that works without any `pyproject.toml`.
